File: src/Microsoft.DotNet.Interactive/Connection/ProxyKernel.cs
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.Events;

namespace Microsoft.DotNet.Interactive.Connection;

public sealed class ProxyKernel : Kernel
{
    private readonly IKernelCommandAndEventSender _sender;
    private readonly IKernelCommandAndEventReceiver _receiver;
    private ExecutionContext _executionContext;

    private readonly Dictionary<string, (KernelCommand command, ExecutionContext executionContext, TaskCompletionSource<KernelEvent> completionSource, KernelInvocationContext
        invocationContext)> _inflight = new();

    public ProxyKernel(
        string name,
        IKernelCommandAndEventSender sender,
        IKernelCommandAndEventReceiver receiver,
        Uri remoteUri = null) : base(name)
    {
        KernelInfo.IsProxy = true;
        _receiver = receiver ?? throw new ArgumentNullException(nameof(receiver));
        _sender = sender ?? throw new ArgumentNullException(nameof(sender));

        if (remoteUri is not null)
        {
            KernelInfo.RemoteUri = remoteUri;
        }
        else if (sender.RemoteHostUri is { } remoteHostUri)
        {
            KernelInfo.RemoteUri = new(remoteHostUri, name);
        }
        else
        {
            throw new ArgumentNullException(nameof(remoteUri));
        }

        var subscription = _receiver.Subscribe(coe =>
        {
            if (coe.Event is { } e)
            {
                if (e is KernelInfoProduced { Command: NoCommand } kip && kip.KernelInfo.Uri == KernelInfo.RemoteUri)
                {
                    UpdateKernelInfoFromEvent(kip);
                    PublishEvent(new KernelInfoProduced(KernelInfo, e.Command));
                }
                else
                {
                    DelegatePublication(e);
                }
            }
        });

        RegisterForDisposal(subscription);
    }

    internal override bool AcceptsUnknownDirectives => true;

    private void UpdateKernelInfoFromEvent(KernelInfoProduced kernelInfoProduced)
    {
        var kernelInfo = kernelInfoProduced.KernelInfo;
        UpdateKernelInfo(kernelInfo);
    }

    private Task HandleByForwardingToRemoteAsync(KernelCommand command, KernelInvocationContext context)
    {
        command.WasProxied = true;

        if (command.OriginUri is null)
        {
            if (context.HandlingKernel == this)
            {
                command.OriginUri = KernelInfo.Uri;
            }
        }

        _executionContext = ExecutionContext.Capture();
        var token = command.GetOrCreateToken();

        command.OriginUri ??= KernelInfo.Uri;

        if (command.DestinationUri is null)
        {
            command.DestinationUri = KernelInfo.RemoteUri;
        }

        if (command is RequestKernelInfo requestKernelInfo)
        {
            if (requestKernelInfo.RoutingSlip.Contains(KernelInfo.RemoteUri, true))
            {
                return Task.CompletedTask;
            }
        }

        var targetKernelName = command.TargetKernelName;
        if (command.TargetKernelName == Name)
        {
            command.TargetKernelName = null;
        }

        var completionSource = new TaskCompletionSource<KernelEvent>();

        var rootToken = KernelCommand.GetRootToken(token);
        _inflight[rootToken] = (command, _executionContext, completionSource, context);

        ExecutionContext.SuppressFlow();

        var t = _sender.SendAsync(command, context.CancellationToken);
        t.ContinueWith(task =>
        {
            if (!task.GetIsCompletedSuccessfully())
            {
                if (task.Exception is { } ex)
                {
                    completionSource.TrySetException(ex);
                }
            }
        });

        return completionSource.Task.ContinueWith(te =>
        {
            command.TargetKernelName = targetKernelName;

            if (te.Result is CommandFailed cf)
            {
                context.Fail(command, cf.Exception, cf.Message);
            }
        });
    }

    private bool CanHandleLocally(KernelCommand command)
    {
        if (!CanHandle(command))
        {
            return false;
        }

        if (HasDynamicHandlerFor(command))
        {
            return true;
        }

        return false;
    }

    protected override Func<TCommand, KernelInvocationContext, Task> CreateDefaultHandlerForCommandType<TCommand>()
    {
        return HandleByForwardingToRemoteAsync;
    }

    internal override Task HandleAsync(
        KernelCommand command,
        KernelInvocationContext context)
    {
        switch (command)
        {
            case AnonymousKernelCommand:
            case DirectiveCommand:
                return base.HandleAsync(command, context);
        }

        if (CanHandleLocally(command))
        {
            return base.HandleAsync(command, context);
        }

        return HandleByForwardingToRemoteAsync(command, context);
    }

    private protected override Task HandleRequestKernelInfoAsync(RequestKernelInfo command, KernelInvocationContext context) =>
        // override the default handler on Kernel and forward the command instead
        HandleByForwardingToRemoteAsync(command, context);

    private void DelegatePublication(KernelEvent kernelEvent)
    {
        var rootToken = KernelCommand.GetRootToken(kernelEvent.Command.GetOrCreateToken());

        var hasPending = _inflight.TryGetValue(rootToken, out var pending);

        var inflightParents = _inflight.Values.Where(v => kernelEvent.Command.IsSelfOrDescendantOf(v.command)).ToArray();

        if (inflightParents.Length == 1)
        {
            pending = inflightParents.Single();
            hasPending = pending.command.HasSameRootCommandAs(kernelEvent.Command);
        }

        if (hasPending && HasSameOrigin(kernelEvent))
        {
            var areSameCommand = pending.command.Equals(kernelEvent.Command);

            if (areSameCommand)
            {
                pending.command.RoutingSlip.ContinueWith(kernelEvent.Command.RoutingSlip);
            }

            switch (kernelEvent)
            {
                case CommandFailed cf when areSameCommand:
                    _inflight.Remove(rootToken);
                    pending.completionSource.TrySetResult(cf);
                    break;
                case CommandSucceeded cs when areSameCommand:
                    _inflight.Remove(rootToken);
                    pending.completionSource.TrySetResult(cs);
                    break;
                case KernelInfoProduced kip when kip.KernelInfo.Uri == KernelInfo.RemoteUri:
                    {
                        UpdateKernelInfoFromEvent(kip);
                        var newEvent = new KernelInfoProduced(KernelInfo, kernelEvent.Command);
                        
                        newEvent.RoutingSlip.ContinueWith(kip.RoutingSlip);

                        if (pending.executionContext is { } ec)
                        {
                            ExecutionContext.Run(ec.CreateCopy(), _ =>
                            {
                                pending.invocationContext.Publish(newEvent);
                                pending.invocationContext.Publish(kip);
                            }, null);
                        }
                        else
                        {
                            pending.invocationContext.Publish(newEvent);
                            pending.invocationContext.Publish(kip);
                        }
                    }
                    break;
                default:
                    {
                        if (pending.executionContext is { } ec)
                        {
                            ExecutionContext.Run(ec.CreateCopy(), _ => pending.invocationContext.Publish(kernelEvent), null);
                        }
                        else
                        {
                            pending.invocationContext.Publish(kernelEvent);
                        }
                    }
                    break;
            }
        }
    }

    private bool HasSameOrigin(KernelEvent kernelEvent)
    {
        var commandOriginUri = kernelEvent.Command.OriginUri;

        if (commandOriginUri is null)
        {
            commandOriginUri = KernelInfo.Uri;
        }

        return commandOriginUri.Equals(KernelInfo.Uri);
    }

    public void UpdateKernelInfo(KernelInfo kernelInfo)
    {
        KernelInfo.IsComposite = kernelInfo.IsComposite;
        KernelInfo.LanguageName = kernelInfo.LanguageName;
        KernelInfo.LanguageVersion = kernelInfo.LanguageVersion;
        KernelInfo.UpdateSupportedKernelCommandsFrom(kernelInfo);
        ((HashSet<KernelDirective>)KernelInfo.SupportedDirectives).UnionWith(kernelInfo.SupportedDirectives);
        if (!string.IsNullOrWhiteSpace(kernelInfo.Description))
        {
            KernelInfo.Description = kernelInfo.Description;
        }
    }
}

File: src/polyglot-notebooks/src/proxyKernel.ts
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import * as commandsAndEvents from "./commandsAndEvents";
import { Logger } from "./logger";
import { Kernel, IKernelCommandHandler, IKernelCommandInvocation, getKernelUri } from "./kernel";
import * as connection from "./connection";
import * as routingSlip from "./routingslip";
import { PromiseCompletionSource } from "./promiseCompletionSource";
import { KernelInvocationContext } from "./kernelInvocationContext";

export class ProxyKernel extends Kernel {

    constructor(override readonly name: string, private readonly _sender: connection.IKernelCommandAndEventSender, private readonly _receiver: connection.IKernelCommandAndEventReceiver, languageName?: string, languageVersion?: string) {
        super(name, languageName, languageVersion);
        this.kernelInfo.isProxy = true;
    }

    override getCommandHandler(commandType: commandsAndEvents.KernelCommandType): IKernelCommandHandler | undefined {
        return {
            commandType,
            handle: (invocation) => {
                return this._commandHandler(invocation);
            }
        };
    }

    private delegatePublication(envelope: commandsAndEvents.KernelEventEnvelope, invocationContext: KernelInvocationContext): void {
        let alreadyBeenSeen = false;
        const kernelUri = getKernelUri(this);
        if (kernelUri && !envelope.routingSlip.contains(kernelUri)) {
            envelope.routingSlip.stamp(kernelUri);
        } else {
            alreadyBeenSeen = true;
        }

        if (this.hasSameOrigin(envelope)) {
            if (!alreadyBeenSeen) {
                invocationContext.publish(envelope);
            }
        }
    }

    private hasSameOrigin(envelope: commandsAndEvents.KernelEventEnvelope): boolean {
        let commandOriginUri = envelope.command?.command?.originUri ?? this.kernelInfo.uri;
        if (commandOriginUri === this.kernelInfo.uri) {
            return true;
        }

        return commandOriginUri === null;
    }

    private updateKernelInfoFromEvent(kernelInfoProduced: commandsAndEvents.KernelInfoProduced) {
        connection.updateKernelInfo(this.kernelInfo, kernelInfoProduced.kernelInfo);
    }

    private async _commandHandler(commandInvocation: IKernelCommandInvocation): Promise<void> {
        const commandToken = commandInvocation.commandEnvelope.getOrCreateToken();
        const commandId = commandInvocation.commandEnvelope.id;
        const completionSource = new PromiseCompletionSource<commandsAndEvents.KernelEventEnvelope>();
        const command = commandInvocation.commandEnvelope;
        // fix : is this the right way? We are trying to avoid forwarding events we just did forward
        let eventSubscription = this._receiver.subscribe({
            next: (envelope) => {
                if (connection.isKernelEventEnvelope(envelope)) {
                    if (envelope.eventType === commandsAndEvents.KernelInfoProducedType &&
                        (envelope.command === null || envelope.command === undefined)) {

                        const kernelInfoProduced = <commandsAndEvents.KernelInfoProduced>envelope.event;
                        kernelInfoProduced.kernelInfo;//?
                        this.kernelInfo;//?
                        if (kernelInfoProduced.kernelInfo.uri === this.kernelInfo.remoteUri) {

                            this.updateKernelInfoFromEvent(kernelInfoProduced);
                            const event = new commandsAndEvents.KernelEventEnvelope(commandsAndEvents.KernelInfoProducedType, { kernelInfo: this.kernelInfo });
                            this.publishEvent(event);
                        }
                    }
                    else if (envelope.command!.getOrCreateToken() === commandToken) {

                        Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] processing event, envelopeid=${envelope.command!.id}, commandid=${commandId}`);
                        Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] processing event, ${JSON.stringify(envelope)}`);

                        try {
                            const original = [...commandInvocation.commandEnvelope?.routingSlip.toArray() ?? []];
                            commandInvocation.commandEnvelope.routingSlip.continueWith(envelope.command!.routingSlip);
                            //envelope.command!.routingSlip = [...commandInvocation.commandEnvelope.routingSlip ?? []];//?
                            Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, command routingSlip :${original}] has changed to: ${JSON.stringify(commandInvocation.commandEnvelope.routingSlip ?? [])}`);
                        } catch (e: any) {
                            Logger.default.error(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, error ${e?.message}`);
                        }

                        switch (envelope.eventType) {
                            case commandsAndEvents.KernelInfoProducedType:
                                {
                                    const kernelInfoProduced = <commandsAndEvents.KernelInfoProduced>envelope.event;
                                    if (kernelInfoProduced.kernelInfo.uri === this.kernelInfo.remoteUri) {
                                        this.updateKernelInfoFromEvent(kernelInfoProduced);
                                        const event = new commandsAndEvents.KernelEventEnvelope(
                                            commandsAndEvents.KernelInfoProducedType,
                                            { kernelInfo: this.kernelInfo },
                                            commandInvocation.commandEnvelope
                                        );

                                        event.routingSlip.continueWith(envelope.routingSlip);

                                        this.delegatePublication(event, commandInvocation.context);
                                        this.delegatePublication(envelope, commandInvocation.context);
                                    } else {
                                        this.delegatePublication(envelope, commandInvocation.context);
                                    }
                                }
                                break;
                            case commandsAndEvents.CommandFailedType:
                            case commandsAndEvents.CommandSucceededType:
                                Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] finished, envelopeid=${envelope.command!.id}, commandid=${commandId}`);
                                if (envelope.command!.id === commandId) {
                                    Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] resolving promise, envelopeid=${envelope.command!.id}, commandid=${commandId}`);
                                    completionSource.resolve(envelope);
                                } else {
                                    Logger.default.info(`proxy name=${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] not resolving promise, envelopeid=${envelope.command!.id}, commandid=${commandId}`);
                                    this.delegatePublication(envelope, commandInvocation.context);
                                }
                                break;
                            default:
                                this.delegatePublication(envelope, commandInvocation.context);
                                break;
                        }
                    }
                }
            }
        });

        try {
            if (!commandInvocation.commandEnvelope.command.destinationUri || !commandInvocation.commandEnvelope.command.originUri) {
                commandInvocation.commandEnvelope.command.originUri ??= this.kernelInfo.uri;
                commandInvocation.commandEnvelope.command.destinationUri ??= this.kernelInfo.remoteUri;
            }

            commandInvocation.commandEnvelope.routingSlip;//?

            if (commandInvocation.commandEnvelope.commandType === commandsAndEvents.RequestKernelInfoType) {
                const destinationUri = this.kernelInfo.remoteUri!;
                if (commandInvocation.commandEnvelope.routingSlip.contains(destinationUri, true)) {
                    return Promise.resolve();
                }
            }
            Logger.default.info(`proxy ${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] forwarding command ${commandInvocation.commandEnvelope.commandType} to ${commandInvocation.commandEnvelope.command.destinationUri}`);
            this._sender.send(commandInvocation.commandEnvelope);
            Logger.default.info(`proxy ${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] about to await with token ${commandToken} and  commandid ${commandId}`);
            const enventEnvelope = await completionSource.promise;
            if (enventEnvelope.eventType === commandsAndEvents.CommandFailedType) {
                commandInvocation.context.fail((<commandsAndEvents.CommandFailed>enventEnvelope.event).message);
            }
            Logger.default.info(`proxy ${this.name}[local uri:${this.kernelInfo.uri}, remote uri:${this.kernelInfo.remoteUri}] done awaiting with token ${commandToken}} and  commandid ${commandId}`);
        }
        catch (e) {
            commandInvocation.context.fail((<any>e).message);
        }
        finally {
            eventSubscription.unsubscribe();
        }
    }
}
